apps/emqx_cluster_link/src/emqx_cluster_link_routerepl.erl

```diff
@@ -0,0 +1,621 @@
+    ok = stop_link_client(St),
+    keep_state_and_data;
+handshaking(info, {'EXIT', ClientPid, Reason}, St = #st{client = ClientPid}) ->
+    enter_disconnected(Reason, St);
+handshaking(Type, Event, St) ->
+    handle_event(Type, Event, St).
+
+enter_bootstrap(_NeedBootstrap, St = #st{bootstrapped = false}) ->
+    enter_bootstrapping(St);
+enter_bootstrap(_NeedBootstrap = true, St = #st{bootstrapped = true}) ->
+    enter_bootstrapping(St);
+enter_bootstrap(_NeedBootstrap = false, St = #st{bootstrapped = true}) ->
+    enter_bootstrapped(St).
+
+enter_bootstrapped(St = #st{actor = Actor}) ->
+    %% Bootstrap is complete.
+    %% Activate route syncer: route ops accumulated since the start of bootstrap
+    %% will start flowing across MQTT channel.
+    ok = activate_syncer(target_cluster(St), Actor),
+    enter_online(St#st{bootstrapped = true}).
+
+enter_online(St) ->
+    {next_state, online, schedule_heartbeat(St)}.
+
+online(info, {timeout, _TRef, heartbeat}, St = #st{}) ->
+    ok = heartbeat(St),
+    {keep_state, schedule_heartbeat(St#st{heartbeat_timer = undefined})};
+online(info, {'EXIT', ClientPid, Reason}, St = #st{client = ClientPid, actor = Actor}) ->
+    %% Occasional client disconnects are expected.
+    %% Suspend route syncer: it will start accumulating route ops in the state.
+    %% Assuming this is transient condition. otherwise route syncer heap risks
+    %% growing indefinitely.
+    ok = suspend_syncer(target_cluster(St), Actor),
+    enter_disconnected(Reason, cancel_heartbeat(St));
+online(Type, Event, St) ->
+    handle_event(Type, Event, St).
+
+schedule_heartbeat(St = #st{heartbeat_timer = undefined}) ->
+    Interval = emqx_cluster_link_config:actor_heartbeat_interval(),
+    TRef = erlang:start_timer(Interval, self(), heartbeat),
+    St#st{heartbeat_timer = TRef}.
+
+cancel_heartbeat(St = #st{heartbeat_timer = undefined}) ->
+    St;
+cancel_heartbeat(St = #st{heartbeat_timer = TRef}) ->
+    ok = emqx_utils:cancel_timer(TRef),
+    St#st{heartbeat_timer = undefined}.
+
+heartbeat(#st{client = ClientPid, actor = Actor, incarnation = Incarnation}) ->
+    publish_heartbeat(ClientPid, Actor, Incarnation).
+
+handle_disconnect(RC, St = #st{actor = Actor}) ->
+    ?tp(info, "cluster_link_connection_disconnect", #{
+        reason => emqx_reason_codes:name(RC),
+        target_cluster => target_cluster(St),
+        actor => Actor
+    }),
+    ok = stop_link_client(St),
+    keep_state_and_data.
+
+enter_disconnected(Reason, St = #st{actor = Actor, reconnect_at = ReconnectAt}) ->
+    case Reason of
+        %% Emit only debug message if stopped manually.
+        normal -> Level = debug;
+        %% Otherwise info message since interruptions are expected.
+        _ -> Level = info
+    end,
+    ?tp(Level, "cluster_link_connection_down", #{
+        reason => Reason,
+        target_cluster => target_cluster(St),
+        actor => Actor,
+        reconnect_at => ReconnectAt
```

Comment:
this epoch timestamp in log is likely not very useful.
maybe calculate `reconnect_in_ms => Delay`